contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpServerHandler.java

```diff
@@ -0,0 +1,101 @@
+
+/**
+ * This helper class provides routine methods to instrument HTTP servers.
+ *
+ * @param <Q> the HTTP request entity.
+ * @param <P> the HTTP response entity.
+ * @since 0.13
+ */
+public class HttpServerHandler<Q, P> extends HttpHandler<Q, P> {
+
+  /**
+   * Creates a {@link HttpServerHandler} with given parameters.
+   *
+   * @param tracer the Open Census tracing component.
+   * @param extractor the {@code HttpExtractor} used to extract information from the
+   *     request/response.
+   * @param customizer the {@link HttpSpanCustomizer} used to customize span behaviors.
+   * @since 0.13
+   */
+  public HttpServerHandler(
+      Tracer tracer, HttpExtractor<Q, P> extractor, HttpSpanCustomizer<Q, P> customizer) {
+    super(tracer, extractor, customizer);
+  }
+
+  /**
+   * Instrument an incoming request before it is handled. Users should optionally invoke {@link
+   * #handleMessageReceived} after the request is received.
+   *
+   * <p>This method will create a span under the deserialized propagated parent context. If the
+   * parent context is not present, the span will be created under the current context.
+   *
+   * <p>This method will invoke {@link HttpSpanCustomizer#customizeSpanStart} after the span is
+   * successfully created.
+   *
+   * <p>The generated span will NOT be set as current context. User can use the returned value to
+   * control when to enter the scope of this span.
+   *
+   * @param <C> the type of the carrier
+   * @param textFormat the {@code TextFormat} used in HTTP propagation.
+   * @param getter the getter used when extracting information from the {@code carrier}.
+   * @param carrier the entity that holds the HTTP information.
+   * @param request the request entity.
+   * @return a span that represents the response handling process.
+   * @since 0.13
+   */
+  public <C> Span handleStart(
+      TextFormat textFormat, TextFormat.Getter<C> getter, C carrier, Q request) {
+    checkNotNull(textFormat, "textFormat");
+    checkNotNull(getter, "getter");
+    checkNotNull(carrier, "carrier");
+    checkNotNull(request, "request");
+    SpanBuilder spanBuilder = null;
+    String spanName = customizer.getSpanName(request, extractor);
+    String parseError = null;
+    // de-serialize the context
+    try {
+      SpanContext spanContext = textFormat.extract(carrier, getter);
+      spanBuilder = tracer.spanBuilderWithRemoteParent(spanName, spanContext);
+    } catch (Exception e) {
+      // record this exception
+      spanBuilder = tracer.spanBuilder(spanName);
+      parseError = e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage();
+    }
+
+    customizer.customizeSpanBuilder(request, spanBuilder, extractor);
+    Span span = spanBuilder.startSpan();
+    // log an annotation to indicate the error
+    if (parseError != null) {
+      span.addAnnotation(Annotation.fromDescription("Error parsing span context: " + parseError));
```

Comment:
Should this really be an annotation on the span and just a simple info-log? Having a client not sending a span context is totally normal and expected. Especially when the service is the "first on in the row".

I would feel weird to have the same Annotation as an "error" (in fact it's not an error, but expected) on each span.

Reply:
Removed.